File: src/countingInversions/countingInversions.java
package countingInversions;

import java.util.Arrays;

import static java.lang.Math.ceil;

class ReturnValues {
    int numberOfInversions;
    int[] sortedArray;
    public ReturnValues(int numberOfInversions, int[] sortedArray) {
        this.numberOfInversions = numberOfInversions;
        this.sortedArray = sortedArray;
    }
}

public class countingInversions {
    public static ReturnValues countInversions(int[] intArray) {
        int arrayLength = intArray.length;
        int numberOfInversions = 0;
        if (arrayLength == 1) {
            return new ReturnValues(0, intArray);
        }
        int midPoint = (int) ceil(arrayLength / 2);
        int[] leftSplit = Arrays.copyOfRange(intArray, 0, midPoint);
        int[] rightSplit = Arrays.copyOfRange(intArray, midPoint, arrayLength);
        ReturnValues sortedLeftSplit = countInversions(leftSplit);
        numberOfInversions += sortedLeftSplit.numberOfInversions;
        ReturnValues sortedRightSplit = countInversions(rightSplit);
        numberOfInversions += sortedRightSplit.numberOfInversions;
        return mergeSplits(numberOfInversions, sortedLeftSplit.sortedArray, sortedRightSplit.sortedArray);
    }

    public static ReturnValues mergeSplits(int numberOfInversions, int[] leftSplit, int[] rightSplit) {
        int l = 0;
        int r = 0;
        int leftLength = leftSplit.length;
        int rightLength = rightSplit.length;
        int totalLength = leftLength + rightLength;
        int[] mergedArrays = new int[totalLength];
        for (int k = 0; k < totalLength; k++) {
            if (l != leftLength && r == rightLength) {
                mergedArrays[k] = leftSplit[l];
                l++;
            }
            else if (l == leftLength && r != rightLength) {
                mergedArrays[k] = rightSplit[r];
                r++;
            }
            else if (leftSplit[l] <= rightSplit[r]) {
                mergedArrays[k] = leftSplit[l];
                l++;
            }
            else if (leftSplit[l] > rightSplit[r]) {
                mergedArrays[k] = rightSplit[r];
                r++;
                numberOfInversions += leftLength - l;
            }
        }
        return new ReturnValues(numberOfInversions, mergedArrays);
    }

    public static void main(String[] args) {
        int[] intArray = {25, 87, 2, 96, 10, 44, 1, 3, 77, 64};
//        int[] intArray = {1, 3, 5, 2, 4, 6};
        ReturnValues sortedInts = countInversions(intArray);
        System.out.println("Given Array: " + Arrays.toString(intArray));
        System.out.println("Sorted Array: " + Arrays.toString(sortedInts.sortedArray));
        System.out.println("Number Of Inversions: " + sortedInts.numberOfInversions);
    }
}

File: src/countingInversions/counting_inversions.py
import math


def countInversions(int_array):
    array_length = len(int_array)
    number_of_inversions = 0
    if array_length == 1:
        return (0, int_array)
    mid_point = math.floor(array_length / 2)
    left_split = int_array[:mid_point]
    right_split = int_array[mid_point:]
    left_inversions, sorted_left_split = countInversions(left_split)
    number_of_inversions += left_inversions
    right_inversions, sorted_right_split = countInversions(right_split)
    number_of_inversions += right_inversions
    return merge_splits(number_of_inversions, sorted_left_split, sorted_right_split)


def merge_splits(number_of_inversions, left_split, right_split):
    l = 0
    r = 0
    k = 0
    left_length = len(left_split)
    right_length = len(right_split)
    total_length = left_length + right_length
    sorted_array = [None] * total_length
    while k < total_length:
        if l != left_length and r == right_length:
            sorted_array[k] = left_split[l]
            l += 1
        elif l == left_length and r != right_length:
            sorted_array[k] = right_split[r]
            r += 1
        elif left_split[l] <= right_split[r]:
            sorted_array[k] = left_split[l]
            l += 1
        elif left_split[l] > right_split[r]:
            sorted_array[k] = right_split[r]
            r += 1
            number_of_inversions += left_length - l
        k += 1
    return (number_of_inversions, sorted_array)


def main(input_array):
    number_of_inversions, sorted_array = countInversions(input_array)
    print("Given Array: ", input_array)
    print("Sorted Array: ", sorted_array)
    print("Number Of Inversions: ", number_of_inversions)

if __name__ == "__main__":
    int_array = [25, 87, 2, 96, 10, 44, 1, 3, 77, 64]
    # int_array = [1, 3, 5, 2, 4, 6]
    main(int_array)
